qiskit/qasm3/grammar.py

```diff
@@ -0,0 +1,644 @@
+
+    def qasm(self):
+        """Unparses the node"""
+        raise NotImplementedError(self)
+
+
+class Statement(Class):
+    """
+    statement
+        : expressionStatement
+        | assignmentStatement
+        | classicalDeclarationStatement
+        | branchingStatement
+        | loopStatement
+        | endStatement
+        | aliasStatement
+        | quantumStatement
+    """
+
+    pass
+
+
+class Pragma(Class):
+    """
+    pragma
+        : '#pragma' LBRACE statement* RBRACE  // match any valid openqasm statement
+    """
+
+    def __init__(self, content):
+        self.content = content
+
+    def qasm(self):
+        return [f"#pragma {self.content};\n"]
+
+
+class CalibrationGrammarDeclaration(Statement):
+    """
+    calibrationGrammarDeclaration
+        : 'defcalgrammar' calibrationGrammar SEMICOLON
+    """
+
+    def __init__(self, name):
+        self.name = name
+
+    def qasm(self):
+        return [f"defcalgrammar {self.name.qasm()};\n"]
+
+
+class Program(Class):
+    """
+    program
+        : header (globalStatement | statement)*
+    """
+
+    def __init__(self, header, statements=None):
+        self.header = header
+        self.statements = statements or []
+
+    def qasm(self):
+        ret = self.header.qasm()
+        for statement in self.statements:
+            if isinstance(statement, str):
+                ret.append(statement)
+            else:
+                ret.append(statement.qasm())
+        return ret
+
+
+class Header(Class):
+    """
+    header
+        : version? include*
+    """
+
+    def __init__(self, version, includes):
+        self.version = version
+        self.includes = includes
+
+    def qasm(self):
+        ret = [self.version.qasm()]
+        for include in self.includes:
+            ret.append(include.qasm())
+        return ret
+
+
+class Include(Class):
+    """
+    include
+        : 'include' StringLiteral SEMICOLON
+    """
+
+    def __init__(self, filename):
+        self.filename = filename
+
+    def qasm(self):
+        return [f"include {self.filename};\n"]
+
+
+class Version(Class):
+    """
+    version
+        : 'OPENQASM'(Integer | RealNumber) SEMICOLON
+    """
+
+    def __init__(self, version_number):
+        self.version_number = version_number
+
+    def qasm(self):
+        return [f"OPENQASM {self.version_number};\n"]
+
+
+class QuantumInstruction(Class):
+    """
+    quantumInstruction
+        : quantumGateCall
+        | quantumPhase
+        | quantumMeasurement
+        | quantumReset
+        | quantumBarrier
+    """
+
+    def __init__(self):
+        pass
+
+    def qasm(self):
+        raise NotImplementedError
+
+
+class Identifier(Class):
+    """
+    Identifier : FirstIdCharacter GeneralIdCharacter* ;
+    """
+
+    def __init__(self, string):
+        self.string = string
+
+    def qasm(self):
+        return self.string
+
+
+class PhysicalQubitIdentifier(Identifier):
+    """
+    TOOD
```

Comment:
Is this a TODO because they haven't been added to the grammar yet, or something else?

Reply:
They are not a grammar category indeed. The `TODO` is on the docstring.